### ISSUE TITLE
"Video source is stereo" tickbox causes black screen

### ISSUE THREAD

**Issue by Axelius** [artoolkit/arunity5#38]
Hello,

we're using ARToolkit 5.3.1 in Unity Personal 5.2.3f1 64bit to create a Video See Through HMD application. Our hardware setup consists of two USB cameras attached to an Oculus Rift DK2. Our platform is Windows 8.1 64bit for now as a prototype for experiments, but in the long run we might target Android.

Unfortunately for me as soon as I check the tickbox "video source is stereo" of the AR Controller no video is rendered to the screen any more (in fact the textures of both video source textures are just black) and there is also no tracking (I thought maybe tracking is still active, but somehow the video is just not rendered to the texture, but this is not the case). It does not even work for the stereo demo scenes provided in the package. When disabling "video source is stereo" it works fine, but of course has only monocular video.

I saw in [this post](https://github.com/artoolkit/artoolkit5/issues/14) of the artoolkit5 issue tracker that maybe a check in VideoSource::updateTexture in ARWrapper.dll might be causing an issue, but this check is never triggered. In my case VideoSource::updateTexture32 seems to be used anyway, but changing the equivalent line in this function also does nothing. When "video source is stereo" is enabled these functions seem not to be called at all.

Just to make sure that ARToolkit5 can at least access both of my cameras at the same time I tried out stereo.exe and it can show video from both cameras.

Can anybody confirm if this feature is broken or am I just missing something? I would really love to use stereo cameras.

Thank you!

Alex

**Comment by wallywyyoung** [comment]
Hello Alex! Have you defined video configuration strings for both the left and right camera? You configure this on the ARController object. Available flags are shown in this doc:
http://artoolkit.org/documentation/doku.php?id=2_Configuration:config_video_capture

**Comment by Axelius** [comment]
My video config strings that I set in Unity are
Video config. (Windows): "-showDialog -flipV"
Video config. (R) (Windows) "-devNum=2 -showDialog -flipV"
I think these are the default values for these fields. And I'm pretty sure that the devNum is correct as the configuration dialog shows up twice and the settings look as if they're corresponding to my cameras.

**Comment by wallywyyoung** [comment]
Do you have any log output you could include?

**Comment by Axelius** [comment]
You mean the Unity Editor Log? I've attached it.

Thanks for looking into this!
[Editor.log.txt](https://github.com/artoolkit/arunity5/files/64747/Editor.log.txt)

**Comment by Axelius** [comment]
Further investigation of this problem resulted in the following:
-Problem persists on another pc with different cameras attached (ARToolkit 5.3.1 in Unity Personal 5.2.1f1 64bit on Win 8.1 64 bit in my case)
-ARToolKit in Unity seems to be able to establish connection only to the first camera as evident by the fact that the LED on the camera turns on only for the first camera
-Problem persists when changing the Graphics API from Direct3D11 to OpenGLCore

As you see I'm really in the dark here. Tomorrow I might try older versions of ARToolKit and/or older versions of Unity to see if there is any change...

**Comment by wallywyyoung** [comment]
@Axelius , have you tried different devNum(s)? I wouldn't doubt it's 0-indexed. Those flags are applicable for -device=WinDS, so try prepending the stream with that, to ensure we're using the correct input pathway.

**Comment by Axelius** [comment]
Yes, if tried different devNum(s). If I exchange -devNum=1 with -devNum=2 **monoscopic** video is definitely working in Unity and the other camera is used. Switching to stereo results again in a black screen.

If I try -devNum=0 or 3,4,... it results in an Error message in my unity viewport
> Unable to start AR tracking. Please check that you have a camera connected.

and in the Unity console it says:
> ARController (native): [error]ARWrap::ARToolKitVideoSource::open(): arVideoOpen unable to open connection to camera using configuration '-device=WinDS -devNum=0 -flipV', exiting returning false
> ARController (native): [error]AR_LOG_LEVEL_ERROR, ARController::startRunning(): Error: unable to open video source 1, exiting, returning false

If I start the stereo.exe in my ARToolKit5 folder with
> C:\Program Files (x86)\ARToolKit5\bin>stereo.exe --vconfL "-device=WinDS -devNum=2 -flipV" --vconfR "-device=WinDS -devNum=1 -flipV"

it works and shows both camera views simultaneously. (These are the same video config strings that I'm also using in Unity now)

I also tried the software iSpy (not trying to promote anything) and it can also show both webcam streams at the same time.

Before including ARToolKit in Unity I had two WebCamTextures rendering the webcam stream to two planes in my scene so I'm sure it's possible to access my two cameras in Unity.

**Comment by Axelius** [comment]
I may have found some kind of **workaround**:
I have set "Video config. (Windows)" to:
> -device=WinDS -devNum=2 -flipV

and "Video config.(R) (Windows)" to 
```xml
-device=WinDSVL <?xml version="1.0" encoding="UTF-8"?><dsvl_input><camera show_format_dialog="false" friendly_name="HD USB Camera"><pixel_format><RGB32 flip_h="false" flip_v="true"/></pixel_format></camera></dsvl_input>
```
Now both videos are shown. (If anybody in the future wants to copy paste this make sure to change your friendly_name string)
Now my left video uses 1920x1080@3Bpp (AR_PIXEL_FORMAT_BGR) and right 1920x1080@4Bpp (AR_PIXEL_FORMAT_BGRA), maybe this is the decisive difference that makes it work? Or maybe for some reason only one device can be used with WinDS and one with WinDSVL in my setup?

Right now I come only one step further: Both videos run fine, but once I move the "Hiro" marker in view of the cameras, the complete Unity Editor crashes... I will look further into this. If I think this is a completely unrelated bug I can't find a solution for, I'll post a new issue.

**Comment by wallywyyoung** [comment]
Thank you, @Axelius. This is definitely strange. Let me check with the team and see if anything like this is a known bug.

**Comment by hanseuljun** [comment]
I'm now facing the same problem that Axelius have met while his solution does not work in my case.

I have figured out an interesting fact from https://github.com/artoolkit/arunity5/blob/master/src/Unity/Assets/ARToolKit5-Unity/Scripts/ARController.cs#L751

Before applying Axelius' solution, the "PluginFunctions.arwCapture()" always returned false but after applying it, it returns true.
Maybe this difference solved the problem for Axelius but mine is still black.
I would be glad to know further about this problem.
